core/headerchain.go

```diff
@@ -591,9 +609,9 @@ func (hc *HeaderChain) SetHead(head uint64, updateFn UpdateHeadBlocksCallback, d
 		markerBatch := hc.chainDb.NewBatch()
 		if updateFn != nil {
 			newHead, force := updateFn(markerBatch, parent)
-			if force && newHead < head {
-				log.Warn("Force rewinding till ancient limit", "head", newHead)
-				head = newHead
+			if force && ((headTime > 0 && newHead.Time < headTime) || (headTime == 0 && newHead.Number.Uint64() < headBlock)) {
+				log.Warn("Force rewinding till ancient limit", "head", newHead.Number.Uint64())
+				headBlock, headTime = newHead.Number.Uint64(), 0
```

Comment:
Perhaps

`headBlock, headTime = newHead.Number.Uint64(), newHead.Time`
to make two conditions aligned?

Reply:
No, we always either want to revert based on timestamps or blocks. What this statement does it that it switches the timestamp based sethead to block based (since we already passed the timestamp).

Reply:
Added a comment to highlight this explicit thought